cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/common/TransformExecutor.java

```diff
@@ -68,73 +56,71 @@ public TransformResponse runOneIteration(IN input) throws Exception {
         // terminal node
-        terminalNodeEntriesMap.put(key, emitterEntries.get(key));
+        List<Object> entries =
+          transformDetailEntry.getValue().getDefaultEmitter().getEntries(transformDetailEntry.getKey());
+        if (entries != null) {
+          terminalNodeEntriesMap.put(transformDetailEntry.getKey(), entries);
+        }
       }
     }
 
-    Map<String, List<InvalidEntry<Object>>> errors = defaultEmitter.getErrors();
+    Map<String, List<InvalidEntry<Object>>> errors =  transformDetailMap.get(start).getDefaultEmitter().getErrors();
     return new TransformResponse(terminalNodeEntriesMap, errors);
   }
 
   private <T> void executeTransformation(final String stageName, List<T> input) throws Exception {
-    Transformation<T, Object> transformation = trackedTransformations.get(stageName);
+    TransformDetail transformDetail = transformDetailMap.get(stageName);
+    Transformation<T, Object> transformation = transformDetail.getTransformation();
 
     // clear old data for this stageName if its not a terminal node
     if (input == null) {
       return;
     }
 
-    if (connectionsMap.containsKey(stageName) && defaultEmitter.getEntriesMap().containsKey(stageName)) {
+    if (!transformDetail.getNextStages().isEmpty() &&
+      transformDetail.getDefaultEmitter().getEntriesMap().containsKey(stageName)) {
       // clear old data if this node was used in a different path earlier during execution.
-      defaultEmitter.getEntries(stageName).clear();
+      transformDetail.getDefaultEmitter().getEntries(stageName).clear();
     }
 
-    if (trackedTransformations.containsKey(stageName)) {
+    if (transformDetail.getTransformation() != null) {
       // has transformation (could be source or transform or sink)
       for (T inputEntry : input) {
-        transformation.transform(inputEntry, new Emitter<Object>() {
-          @Override
-          public void emit(Object value) {
-            defaultEmitter.emit(stageName, value);
-          }
-
-          @Override
-          public void emitError(InvalidEntry<Object> invalidEntry) {
-            defaultEmitter.emitError(stageName, invalidEntry);
-          }
-        });
+        transformation.transform(inputEntry, transformDetail);
       }
     }
 
-    List<String> nextStages = connectionsMap.get(stageName);
-    if (nextStages != null) {
+    List<String> nextStages = transformDetail.getNextStages();
+    if (!nextStages.isEmpty()) {
       // transform or source
       for (String nextStage : nextStages) {
-        executeTransformation(nextStage, defaultEmitter.getEntries(stageName));
+        executeTransformation(nextStage, transformDetail.getDefaultEmitter().getEntries(stageName));
       }
     } else {
       // terminal node, pass on the input if transformation for this terminal node is not already executed.
-      if (!trackedTransformations.containsKey(stageName)) {
+      if (transformDetail.getTransformation() == null) {
```

Comment:
I see now this was for the realtime case.  We shouldn't have special logic like this, it couples the transform executor with the realtime worker. This class should deal only with transforms, which is possible if you pass in a list of first stages in the constructor rather than just a single first stage.